### PR TITLE
 nixos-rebuild: fix issues when using `--target-host`

### PR DESCRIPTION
**Pull request by Enzime**
## Description of changes

This PR fixes deploying to hosts that require entering a password for `sudo` using `--target-host`.

This PR changes `--use-remote-sudo` to use `sudo` as little as possible.

This PR also fixes `systemd-run` hanging when used with `--target-host`.

Replaces #109046

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @FRidh @zimbatm @oxalica @bjornfor 

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Enzime** [comment]
cc @thiagokokada @duament @roberth

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh [hunk 1]
```diff
@@ -157,8 +161,14 @@ while [ "$#" -gt 0 ]; do
...
-if [[ -n "$SUDO_USER" || -n $remoteSudo ]]; then
-    maybeSudo=(sudo --preserve-env="$preservedSudoVars" --)
+sudoCommand=(sudo --preserve-env="$preservedSudoVars" --)
+
+if [[ -n "$SUDO_USER" ]]; then
+    maybeSudo=("${sudoCommand[@]}")
+fi
+
+if [[ -n "$remoteSudoAlways" ]]; then
+    maybeSudo=(-t "${sudoComand[@]}")
```
Comment (roberth) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:171]: This seems to rely on a coincidence that `maybeSudo` gets pasted onto some other command consistently.
Please don't do that.

**Review by roberth — CHANGES_REQUESTED** [review]
Unfortunately, if you don't add a test, it will keep breaking. Please add a test.

**Comment by duament** [comment]
I tried this pr. But it didn't fix #262686.

My first switch (https://github.com/duament/flakes/actions/runs/7363818666/job/20043632674) succeeded, although there's a warning `Pseudo-terminal will not be allocated because stdin is not a terminal.` and the output of switch-to-configuration was not printed.

My second switch (https://github.com/duament/flakes/actions/runs/7363898397/job/20043786127) failed, but the exit code was 0 and github treated it as a success. Here's the error message: `Failed to start transient service unit: Unit nixos-rebuild-switch-to-configuration.service was already loaded or has a fragment file.`

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh (hunk 1 above)
Comment (Enzime) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh]: I've changed it to always use `-t` with `ssh` as a PTY may also be necessary with non-`sudo` commands e.g. connecting for the first time and SSH asking if you want to trust an unknown host

**Comment by Enzime** [comment]
@duament could you test this PR again? I've changed the flag from `-t` to `-tt` which should hopefully fix it

**Comment by Enzime** [comment]
`-tt` breaks the tests so I'm going to switch back to using `-t`, my last commit only fixes a `systemd-run` hang regression that was introduced by adding `-t`, I don't think it fixes #262686

File: nixos/tests/nixos-rebuild-target-host.nix [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  targetAddress = "192.168.0.2";
+in {
+  name = "nixos-rebuild-target-host";
+
+  nodes = {
+    deployer = { lib, ... }: {
+      imports = [ ../modules/profiles/installation-device.nix ];
+
+      networking.interfaces.eth1.ipv4.addresses = [ { address = deployerAddress; prefixLength = 24; } ];
+      networking.hosts.${targetAddress} = [ "target" ];
```
Comment (roberth) [inline nixos/tests/nixos-rebuild-target-host.nix:14]: I though this happened automatically. Can this line be removed?

File: nixos/tests/nixos-rebuild-target-host.nix [hunk 3]
```diff
@@ -0,0 +1,96 @@
+import ./make-test-python.nix ({ pkgs, ... }: let
+  inherit (import ./ssh-keys.nix pkgs) snakeOilPrivateKey snakeOilPublicKey;
+
+  deployerAddress = "192.168.0.1";
+  targetAddress = "192.168.0.2";
```
Comment (roberth) [inline nixos/tests/nixos-rebuild-target-host.nix:5]: You could get this from the `nodes` module argument instead.
It's available at the root of the test, as well as in the individual configs, and as an argument to `testScript`.
Not sure if `make-test-python.nix` exposes module functionality though. Maybe switch from `handleTest` to `runTest` if you want to keep using a near-top-level `let`?

File: nixos/tests/nixos-rebuild-target-host.nix [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+          boot.loader.grub = {
+            enable = true;
+            device = "/dev/vda";
+            forceInstall = true;
+          };
+
+          documentation.enable = false;
+
+          networking.hostName = "deployed";
```
Comment (roberth) [inline nixos/tests/nixos-rebuild-target-host.nix:72]:
```suggestion
          # this will be asserted
          networking.hostName = "deployed";
```

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.8 [hunk 5]
```diff
@@ -30,6 +30,7 @@
 .Op Fl -build-host Va host
 .Op Fl -target-host Va host
 .Op Fl -use-remote-sudo
+.Op Fl -use-remote-sudo-always
```
Comment (roberth) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.8:33]: If we really need a new behavior (do we?), maybe fix the inconsistency of "remote" vs "target"?
Not sure I like "target" because of cross compilation terminology, but target does become less relevant to users as compilers and their Nix expressions improve. "Remote" isn't great either, because both machines may be equally remote, as demonstrated by the test.
Also if you have an office with NixOS machines and you make the on site server responsible for deploying updates, arguably the deployer is more remote.

So I think I prefer target?

```suggestion
.Op Fl -target-use-sudo
```

Using `target` as a prefix reinforces nice grouping, and helps with discovery through shell completions.

File: nixos/tests/nixos-rebuild-target-host.nix (hunk 2 above)
Comment (Enzime) [reply nixos/tests/nixos-rebuild-target-host.nix]: Done :+1:

File: nixos/tests/nixos-rebuild-target-host.nix (hunk 3 above)
Comment (Enzime) [reply nixos/tests/nixos-rebuild-target-host.nix]: Done :+1:

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.8 (hunk 5 above)
Comment (Enzime) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.8]: I've removed the `--use-remote-sudo-always` flag, opting to just make this PR an improvement on the existing `--use-remote-sudo` flag

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh [hunk 6]
```diff
@@ -192,6 +194,23 @@ targetHostCmd() {
...
+    fi
+
+    targetHostCmd "$@"
+    ret=$?
+
+    if [[ -n "$restoreMaybeSudo" ]]; then
+        maybeSudo=("${previousMaybeSudo[@]}")
+    fi
+    return $ret
+}
```
Comment (roberth) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:212]: This function is really hard to follow.
I think it's ok to duplicate the implementation of `targetHostCmd`. It's only five lines, and it'd be so much easier to understand that it's also easier to change coherently.

Comment (Enzime) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:203]: I've rewritten the function to be a lot simpler :+1:

File: nixos/tests/nixos-rebuild-target-host.nix [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+          { groups = [ "wheel" ]; commands = [ { command = "ALL"; } ]; }
+          { users = [ "alice" ]; commands = [ { command = "ALL"; options = [ "NOPASSWD" ]; } ]; }
+        ];
+
+        nix.settings.trusted-users = [ "@wheel" ];
+      };
+    in {
+      imports = [ ./common/user-account.nix ];
+
+      config = targetConfig // {
```
Comment (roberth) [inline nixos/tests/nixos-rebuild-target-host.nix:52]:
```suggestion
      config = lib.mkMerge [
        targetConfig
        {
```

A bit more robust, for when changes are needed later.

**Review by roberth — APPROVED** [review]

File: nixos/tests/nixos-rebuild-target-host.nix (hunk 7 above)
Comment (Enzime) [reply nixos/tests/nixos-rebuild-target-host.nix]: Updated :+1:

**Review by roberth — APPROVED** [review]

**Comment by hannesg** [comment]
Hi @Enzime, Hi @roberth 

I just tested the change with my nixos server and I observed a regression. When using `--build-host` and flakes, nixos-rebuild fails with a weird error. I added +x logging to bash and found that `pathToConfig` ( [here](https://github.com/NixOS/nixpkgs/blob/ea3179cf3b5362811bc40137430864f66c685237/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L677) ) is set to the tty output of ssh which is in my case:

```
'warning: you did not specify '\''--add-root'\''; the result might be removed by the garbage collector
'nix/store/ph132c4ihigcfyi86pwlmn4q12qmwyz7-nixos-system-granogayo-24.05.20240113.e0d20bf
```

All subsequent commands then fail with weird errors. Reverting https://github.com/NixOS/nixpkgs/pull/277642/commits/09fd207cb8a6a1e654cdc9dc49610ebd456fe05d fixes the issue.

**Comment by roberth** [comment]
Found an even worse problem as well, which was already caught by the tests, but not reported well by ofborg. I did check the list, but missed the single grey status this time.

Fixes available in #277642; will merge as soon as tests pass locally.

If any other issues come up, consider reverting both PRs unless trivially fixable.

TODO:
- [ ] Add regression test for `--build-host`. I suggest extending the target host test for this; seems more efficient that way. Maybe rename it to `-remote-hosts`? Keep the existing test case in the testScript as is.

**Comment by roberth** [comment]
Made a start here. Would appreciate some help.
 - #281130
